projects/extension/src/background/index.ts

```diff
@@ -191,6 +202,70 @@ window.uiInterface = {
       }
     }
   },
+  // Based on the chain provided, it returns the default bootnodes of given chain
+  // bootnodes are retrieved from the chainspecs that exist in the extension.
+  getDefaultBootnodes: (chain: string): string[] => {
+    if (chain === "polkadot") return polkadot.bootNodes
+    if (chain === "ksmcc3") return ksmcc3.bootNodes
+    if (chain === "westend2") return westend2.bootNodes
+    if (chain === "rococo_v2_2") return rococo_v2_2.bootNodes
+    return []
+  },
+  // Sends an message through rpc to smoldot in order to validate the given bootnode.
+  // This will return to the caller the response once received from rpc.
+  // Message's id is prefixed with "extension:" in order to be differentiated from the rest
+  // and be able to reply specific message at nextSandboxMessage to the caller.
+  updateBootnode: async (
+    chain: string,
+    bootnode: string,
+    add: boolean,
+  ): Promise<p2pDiscoverResp | undefined> => {
+    if (!add) {
+      // removing from localstorage if exist
+      chrome.storage.local.get(["bootNodes_".concat(chain)], (result) => {
+        const res = [...result["bootNodes_".concat(chain)]]
+        const idx = res.findIndex((a) => a === bootnode)
+        if (idx > -1) {
+          res.splice(idx, 1)
+          chrome.storage.local.set({ ["bootNodes_".concat(chain)]: [...res] })
+        }
+      })
+    } else {
+      // validate and add to localstorage
+      if (manager.state !== "ready") return
+      manager.manager.sandboxMessage(null, {
+        chainId: chain,
+        type: "rpc",
+        origin: "substrate-connect-client",
+        jsonRpcMessage: JSON.stringify({
+          id: "extension:" + manager.manager.nextRpcRqId,
+          jsonrpc: "2.0",
+          method: "sudo_unstable_p2pDiscover",
+          params: [bootnode],
+        }),
+      })
+
+      while (true) {
+        const msg = await manager.manager.nextSandboxMessage(null)
+        if (
+          msg.type === "extension-config" &&
+          msg.origin === "connection-manager"
+        ) {
```

Comment:
This is wrong. If there are any other messages pending, you're just silently discarding them.
The logic of that whole thing should be in `ConnectionManagerWithHealth`.